x-pack/plugins/apm/server/routes/storage_explorer/get_service_statistics.ts

```diff
@@ -0,0 +1,240 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0; you may not use this file except in compliance with the Elastic License
+ * 2.0.
+ */
+import { QueryDslQueryContainer } from '@elastic/elasticsearch/lib/api/typesWithBodyKey';
+import {
+  termQuery,
+  kqlQuery,
+  rangeQuery,
+} from '@kbn/observability-plugin/server';
+import { ProcessorEvent } from '@kbn/observability-plugin/common';
+import { ApmPluginRequestHandlerContext } from '../typings';
+import { Setup } from '../../lib/helpers/setup_request';
+import {
+  IndexLifecyclePhaseSelectOption,
+  indexLifeCyclePhaseToDataTier,
+} from '../../../common/storage_explorer_types';
+import { getTotalTransactionsPerService } from './get_total_transactions_per_service';
+import {
+  PROCESSOR_EVENT,
+  SERVICE_NAME,
+  SERVICE_ENVIRONMENT,
+  TIER,
+  TRANSACTION_SAMPLED,
+  AGENT_NAME,
+  INDEX,
+} from '../../../common/elasticsearch_fieldnames';
+import { environmentQuery } from '../../../common/utils/environment_query';
+import { AgentName } from '../../../typings/es_schemas/ui/fields/agent';
+import {
+  getTotalIndicesStats,
+  getEstimatedSizeForDocumentsInIndex,
+} from './indices_stats_helpers';
+import { RandomSampler } from '../../lib/helpers/get_random_sampler';
+
+async function getMainServiceStatistics({
+  setup,
+  context,
+  indexLifecyclePhase,
+  randomSampler,
+  start,
+  end,
+  environment,
+  kuery,
+}: {
+  setup: Setup;
+  context: ApmPluginRequestHandlerContext;
+  indexLifecyclePhase: IndexLifecyclePhaseSelectOption;
+  randomSampler: RandomSampler;
+  start: number;
+  end: number;
+  environment: string;
+  kuery: string;
+}) {
+  const { apmEventClient } = setup;
+
+  const allIndicesStats = await getTotalIndicesStats({ context, setup });
+
+  const response = await apmEventClient.search('get_main_service_statistics', {
+    apm: {
+      events: [
+        ProcessorEvent.span,
+        ProcessorEvent.transaction,
+        ProcessorEvent.error,
+        ProcessorEvent.metric,
+      ],
+    },
+    body: {
+      size: 0,
+      query: {
+        bool: {
+          filter: [
+            ...environmentQuery(environment),
+            ...kqlQuery(kuery),
+            ...rangeQuery(start, end),
+            ...(indexLifecyclePhase !== IndexLifecyclePhaseSelectOption.All
+              ? termQuery(
+                  TIER,
+                  indexLifeCyclePhaseToDataTier[indexLifecyclePhase]
+                )
+              : []),
+          ] as QueryDslQueryContainer[],
+        },
+      },
+      aggs: {
+        sample: {
+          random_sampler: randomSampler,
+          aggs: {
+            services: {
+              terms: {
+                field: SERVICE_NAME,
+                size: 500,
+              },
+              aggs: {
+                sample: {
+                  top_hits: {
```

Comment:
top_metrics here should help. top_hits is pretty slow with a large number of buckets.

Reply:
Thanks for the feedback, changed it in https://github.com/elastic/kibana/pull/130152/commits/f5f4a6094bbeb82be663042a8d2158ff00bd9d86